### PR TITLE
33294 eliminate aspectmock from allureHelperTest

### PR DESCRIPTION
**Pull request by SilinMykola**
### Description
Eliminate AspectMock usage from `dev/tests/unit/Magento/FunctionalTestFramework/Allure/AllureHelperTest.php`

### Fixed Issues
1. magento/magento2#33294: [MFTF] Eliminate AspectMock from AllureHelperTest (Complex!) #33294

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/verification tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)
 - [ ] Changes to Framework doesn't have backward incompatible changes for tests or have related Pull Request with fixes to tests

### CODE REVIEW

**Comment by jilu1** [comment]
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

**Comment by jilu1** [comment]
@SilinMykola 
Thank you for your pull request! We will review and update you soon.

**Comment by jilu1** [comment]
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

**Comment by jilu1** [comment]
@SilinMykola 
Please fix the 3 unit test failures in this PR?

**Comment by sivaschenko** [comment]
@SilinMykola thanks for the pull request! Would you be able to respond to the code review comments?

**Comment by SilinMykola** [comment]
@sivaschenko hi! Yes, I'll try to fix all tests errors. I need some time for this.

**Comment by bohdan-harniuk** [comment]
Hello, @jilu1

I have a little bit helped for @SilinMykola with the aspect mock replacement here.
Please, proceed with the code review.

cc: @SilinMykola, @sivaschenko

**Review by andrewbess — APPROVED** [review]
Hello @SilinMykola @bohdan-harniuk 
Thank you for your contribution.

**Comment by jilu1** [comment]
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

**Comment by magento-engcom-team** [comment]
@jilu1 the pull request successfully imported.

File: src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php [hunk 1]
```diff
@@ -1,35 +1,63 @@
...
+    ): AddUniqueAttachmentEvent {
+        if (!self::$instance) {
+            self::$instance = new AddUniqueAttachmentEvent(
+                $filePathOrContents,
+                $caption,
+                $type
+            );
+        }
+        return self::$instance;
+    }
```
Comment (jilu1) [inline src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php:46]: This alternative way to instantiate an object of this class is only used for unit tests. This should not be used in MFTF. Am I understand it correctly here? If this is the case, can you make it clear in docblock?

Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php]: It's still confusing to add this static method in this class. Is there any other way?
The particular test is valuable, if this cannot be mocked by PHPUnit, can we turn this into a functional test https://github.com/magento/magento2-functional-testing-framework/tree/develop/dev/tests/functional?

Comment (jilu1) [reply src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php]: If `copyFile()` is the reason that prevents you from mocking the class, you could do something like:
```
private function copyFile($filePath, $outputPath)
{
    if (MftfApplicationConfig::getConfig()->getPhase() == MftfApplicationConfig::UNIT_TEST_PHASE) {
        return true;
    }
    return copy($filePath, $outputPath);
}
```

Comment (bohdan-harniuk) [reply src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php]: Thank you, @jilu1 ! 
It is a great approach for the copyFile method! I will use it for the testAddAttachmentUniqueName testing method!
For other tested methods we have to mock AddUniqueAttachmentEvent with the singleton approach.

Comment (bohdan-harniuk) [reply src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php]: Hello @jilu1!
I've done that.
Thank you for your suggestion!

**Review by bohdan-harniuk — APPROVED** [review]
Hello, @jilu1!

I've fully eliminated the new singleton creation in this PR! It is no blocker anymore!
You can proceed with the code review

Thanks, Bohdan

**Review by jilu1 — CHANGES_REQUESTED** [review]
@bohdan-harniuk 
Have you run this PR with magento `platform-health` branch? I have seen the following exception when running a build:

```
[group329] PHP Fatal error:  Uncaught TypeError: file_exists() expects parameter 1 to be a valid path, object given in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php:34
[group329] Stack trace:
[group329] #0 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php(34): file_exists()
[group329] #1 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Event/AddAttachmentEvent.php(34): Magento\FunctionalTestingFramework\Allure\Event\AddUniqueAttachmentEvent->getAttachmentFileName()
[group329] #2 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(116): Yandex\Allure\Adapter\Event\AddAttachmentEvent->process()
[group329] #3 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(92): Yandex\Allure\Adapter\Allure->processStepEvent()
[group329] #4 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/ in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php on line 34

```
This exception happened in multiple test groups.

**Comment by bohdan-harniuk** [comment]
> @bohdan-harniuk
> Have you run this PR with magento `platform-health` branch? I have seen the following exception when running a build:
> 
> ```
> [group329] PHP Fatal error:  Uncaught TypeError: file_exists() expects parameter 1 to be a valid path, object given in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php:34
> [group329] Stack trace:
> [group329] #0 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php(34): file_exists()
> [group329] #1 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Event/AddAttachmentEvent.php(34): Magento\FunctionalTestingFramework\Allure\Event\AddUniqueAttachmentEvent->getAttachmentFileName()
> [group329] #2 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(116): Yandex\Allure\Adapter\Event\AddAttachmentEvent->process()
> [group329] #3 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(92): Yandex\Allure\Adapter\Allure->processStepEvent()
> [group329] #4 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/ in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php on line 34
> ```
> 
> This exception happened in multiple test groups.

Hello, @jilu1! 
No, I haven’t checked it because we didn't change any business logic here:
Screenshot 2021-07-30 at 17 39 32

Could you please take a look at it? What do you think, how it is possible that we broke something?

Thanks, Bohdan

**Comment by jilu1** [comment]
> > @bohdan-harniuk
> > Have you run this PR with magento `platform-health` branch? I have seen the following exception when running a build:
> > ```
> > [group329] PHP Fatal error:  Uncaught TypeError: file_exists() expects parameter 1 to be a valid path, object given in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php:34
> > [group329] Stack trace:
> > [group329] #0 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php(34): file_exists()
> > [group329] #1 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Event/AddAttachmentEvent.php(34): Magento\FunctionalTestingFramework\Allure\Event\AddUniqueAttachmentEvent->getAttachmentFileName()
> > [group329] #2 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(116): Yandex\Allure\Adapter\Event\AddAttachmentEvent->process()
> > [group329] #3 /var/www/html/vendor/allure-framework/allure-php-api/src/Yandex/Allure/Adapter/Allure.php(92): Yandex\Allure\Adapter\Allure->processStepEvent()
> > [group329] #4 /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/ in /var/www/html/vendor/magento/magento2-functional-testing-framework/src/Magento/FunctionalTestingFramework/Allure/Event/AddUniqueAttachmentEvent.php on line 34
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This exception happened in multiple test groups.
> 
> Hello, @jilu1!
> No, I haven’t checked it because we didn't change any business logic here:
> Screenshot 2021-07-30 at 17 39 32
> 
> Could you please take a look at it? What do you think, how it is possible that we broke something?
> 
> Thanks, Bohdan
@bohdan-harniuk 

I did not see anything obviously also but the failing build is observed when I run 5 PRs (#839, #843, #844, #845, #852) in a batch. Please double-check from your side. I will look into it more.

**Comment by jilu1** [comment]
@SilinMykola @bohdan-harniuk 

FYI: The exception is reproducible running this PR branch alone.

**Comment by karyna-tsymbal-atwix** [comment]
Hi @jilu1  , I've just re-checked our PR. It seems that after the branch was actualized the problem that you described is gone. I don't see any errors on my local environment. Can you please check it one more time?

**Comment by jilu1** [comment]
> Hi @jilu1 , I've just re-checked our PR. It seems that after the branch was actualized the problem that you described is gone. I don't see any errors on my local environment. Can you please check it one more time?

@karyna-tsymbal-atwix 
Have you run this PR with Magento platform-health branch? To reproduce it, you can create a Magento PR based on platform-health and use mftf from this PR. You should see the problem when running the full CE/EE/B2B build. It happens in some test groups but not all of them.

**Comment by bohdan-harniuk** [comment]
Hello, @jilu1 !

I've reproduced mentioned error! I'll fix it!

Thank you!

**Comment by bohdan-harniuk** [comment]
I've successfully reproduced mentioned issue:
Screenshot 2021-08-09 at 17 50 29

My steps to reproduce:

1. Run any MFTF test without running WebDriver (ChromeDriver)
2. Got this error

After fixing this issue I got the correct output:
Screenshot 2021-08-09 at 17 56 28

And successfully finished test:
Screenshot 2021-08-09 at 18 09 40

@jilu1, could you please recheck if everything is fine?

Thank you, 
Bohdan.

**Review by bohdan-harniuk — DISMISSED** [review]

**Comment by jilu1** [comment]
@magento import pull request to https://github.com/magento-commerce/magento2-functional-testing-framework

**Comment by magento-engcom-team** [comment]
@jilu1 the pull request successfully imported.

**Review by andrewbess — APPROVED** [review]

**Pull request merged**